truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumentable.java

```diff
@@ -0,0 +1,86 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.instrumentation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import com.oracle.truffle.api.instrumentation.InstrumentableFactory.WrapperNode;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.source.SourceSection;
+
+/**
+ * <p>
+ * Marks a guest language AST node class as instrumentable. An instrumentable node must provide a
+ * {@link SourceSection}, a set of instrumentation {@link #tags() tags} and a {@link #factory()
+ * wrapper factory} to create a wrapper for the instrumentable node. If a super-class is already
+ * instrumentable and declares a wrapper factory then {@link #factory()} can be left default.
+ * </p>
+ * <p>
+ * {@link Instrumentable} nodes must extend {@link Node}. The instrumentation framework will
+ * {@link Node#replace(Node) replace} the instrumentable node with a {@link WrapperNode} and
+ * delegate to the original node.
+ * </p>
+ * <p>
+ * Example for a minimal implementation of an {@link Instrumentable instrumentable} node with a
+ * generated wrapper.
+ *
+ * {@codesnippet Instrumentable.BaseNode}
+ *
+ * <p>
+ * Example of a subclass not defining a wrapper factory but just tags:
+ * </p>
+ *
+ * {@codesnippet Instrumentable.ExpressionNode}
+ *
+ * @see InstrumentationTag
+ * @see WrapperNode
+ * @see ProbeNode
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE})
+public @interface Instrumentable {
+
+    /**
+     * Assigns a wrapper factory to a {@link Node node} class annotated as {@link Instrumentable
+     * instrumentable}. To use the generated wrapper factory use the is generated with the original
+     * class name and the "Wrapper" suffix. So if the class was called <code>StatementNode</code>
+     * then the generated factory class will be called <code>StatementNodeWraper</code>.
+     *
+     * <pre>
+     * &#064;Instrumentable(factory = BaseNodeWrapper.class)
+     * public abstract class BaseNode extends Node {
+     * 
+     *     public abstract Object execute(VirtualFrame frame);
+     * 
+     * }
+     * </pre>
+     */
+    Class<? extends InstrumentableFactory<? extends Node>> factory() default InheritFactory.class;
```

Comment:
Ok, so there seem to be multiple requirements:
 - class must be non-final
 - execute methods need to be non-final

In addition, it seems to be problematic to have constructors with more than one argument.
I am not entirely sure why that is, I suppose it is a bug in the code generator. I assume that the wrapping node does not need any of the arguments, because the `delegateNode` is already the proper one with everything.